### ISSUE TITLE
Item-Iterator should check hasNext() and not be reused

### ISSUE THREAD

**Issue by d-schmidt** [michael-rapp/Apriori#5]
This was a tricky bug to find:
https://github.com/michael-rapp/Apriori/blob/b03b7cbdc3fb814757844cdd02a035ba5dbfc0a3/src/main/java/de/mrapp/apriori/modules/FrequentItemSetMinerModule.java#L70

My first ugly work-around looks like this:
``` java
        Collection<Transaction<Article>> customers = getCustomers();
        customers.add(null);

        Iterator<Transaction<Article>> iterator = customers.iterator();
        Output<Article> output = apriori.execute(iterator);
```
edit: and this work-around doesn't work. Iterators are also reused, see below

**Comment by michael-rapp** [comment]
Why would you want the iterator to return null? The algorithm cannot handle transactions that are null. At best it can ignore them.

**Comment by d-schmidt** [comment]
## This Bug
I don't want my Iterator to return null, I tried it to to circumvent this wrong use of Iterators.

Iterators never return null on `next()`. The `while` is never false. Iterators throw a `NoSuchElementException` at the end.

The Iterator interface Javadoc:
```java
    /**
     * Returns {@code true} if the iteration has more elements.
     * (In other words, returns {@code true} if {@link #next} would
     * return an element rather than throwing an exception.)
     *
     * @return {@code true} if the iteration has more elements
     */
    boolean hasNext();

    /**
     * Returns the next element in the iteration.
     *
     * @return the next element in the iteration
     * @throws NoSuchElementException if the iteration has no more elements
     */
    E next();
```
You usually use them like this:
```java
    while (iterator.hasNext()) { 
         transaction = iterator.next();
```
Your Iterators are the first implementations I've ever seen to infinitely return `null`. Your code only works because they are irregular. I recommend you change your DataIterator and fix the code afterwards.

## Another Bug
Also you are reusing Iterators. They are made for one-time-use only. You will find this crash when you change your Iterators and run your 2nd example.

This method doesn't work and it should show in a test because it influences the result:
https://github.com/michael-rapp/Apriori/blob/da31097d16844eb28f681092328c044707737e7c/src/main/java/de/mrapp/apriori/tasks/FrequentItemSetMinerTask.java#L84
This line method call throws a NSEE (or in your test implementations is skipped) because the Iterator is already empty:
https://github.com/michael-rapp/Apriori/blob/da31097d16844eb28f681092328c044707737e7c/src/main/java/de/mrapp/apriori/tasks/FrequentItemSetMinerTask.java#L93

## Suggestion
I suggest replacing all Iterators with the Iterable interface. You are able to foreach loop them and they are reusable while still being as open as your current API to allow any kind of data source.

I'm also curious why no tests catches the 2nd bug. The math should be off. I just copied your 2nd example and used a List.iterator() and got the NSEE.

I'm leaving this as tips and will try to find another implementation. These basic mistakes undermine the trust in such an important library.

If you have any further questions, I'm happy to answer them.

**Comment by michael-rapp** [comment]
Of course it's correct that iterators should throw `NoSuchElementException`s instead of returning null. I somehow mixed things up when I was working on this. No need to say, that I will fix this as soon as possible. Nevertheless, it works flawlessly the way it is implemented if I'm not mistaken, although I agree that it's not "best-practice".

I am also aware that iterators are not meant to be reused. I took a quick look at the code and could not identify a piece of code where an iterator is used more than once. However, if that's the case despite my knowledge, I will run into it when working on this issue, I guess. 

I'm sorry, if you lost confidence in this library because of the issues you encountered. However, you should bear in mind, that this was meant to be a fun project and not a professional or even paid job. Mistakes happen and one benefit of open source software is that it gets better the more people care about it (which obviously wasn't the case with this niche library so far). I therefore appreciate your feedback and if you want to contribute, I will consider merging pull requests.

By the way, I wasn't able to find an open source library that is able to search for frequent item sets and to induce association rules at the same time. Especially not a one that provides a neat API. But maybe you are more lucky. At least, there is a Weka implementation for mining frequent item sets as well as some people's personal Github projects.

**Comment by d-schmidt** [comment]
The reuse: Line 93 in FrequentItemSetMinerTask.java is inside a while loop. If the loop is false and it runs again, it uses the same iterator again in findFrequentItemSets().

There are very few libraries and yours looked clean and promising. I wanted to use it in a company project so instead of just dropping it I did the least I could/should do and point out bugs I'd noticed.

Maybe I will come back around and give it another shot.

**Comment by michael-rapp** [comment]
Yes, you're right. An iterable would be the way to go. Nothing that can't be fixed.

**Comment by michael-rapp** [comment]
Solved in version 1.3.0.
